Firestore/core/src/firebase/firestore/remote/grpc_connection.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_GRPC_CONNECTION_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_GRPC_CONNECTION_H_
+
+#include <memory>
+
+#include "Firestore/core/src/firebase/firestore/core/database_info.h"
+#include "Firestore/core/src/firebase/firestore/remote/grpc_stream.h"
+#include "Firestore/core/src/firebase/firestore/remote/grpc_stream_observer.h"
+#include "absl/strings/string_view.h"
+#include "grpcpp/channel.h"
+#include "grpcpp/client_context.h"
+#include "grpcpp/completion_queue.h"
+#include "grpcpp/generic/generic_stub.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+// PORTING NOTE: this class has limited resemblance to `GrpcConnection` in Web
+// client. However, unlike Web client, it's not meant to hide different
+// implementations of a `Connection` under a single interface.
+
+/**
+ * Creates and owns gRPC objects (channel and stub) necessary to produce a
+ * `GrpcStream`.
+ */
+class GrpcConnection {
+ public:
+  GrpcConnection(const core::DatabaseInfo& database_info,
+                 util::AsyncQueue* worker_queue,
+                 grpc::CompletionQueue* grpc_queue);
+
+  /**
+   * Creates a stream to the given stream RPC endpoint. The resulting stream
+   * needs to be `Start`ed before it can be used.
+   */
+  // PORTING NOTE: unlike Web client, the created stream is not open and has to
+  // be started manually.
+  std::unique_ptr<GrpcStream> CreateStream(absl::string_view rpc_name,
+                                           absl::string_view token,
+                                           GrpcStreamObserver* observer);
+
+ private:
+  // PORTING NOTE: this function is called `AddMetadata` in Web client.
```

Comment:
I'm on the fence about flagging this but here goes.

Porting notes should be for things that intended to be durably different across the platforms, especially in cases where they could be expected to be the same but can't be. In this case it seems like either these functions aren't closely related after all (in which case no porting note is necessary) or this should be a TODO to reconcile the differences.

In networking code it seems that due to differences in the APIs we're inevitably stuck with differences and private methods to support those seem inevitable. I don't know that we need to flag them all.